src/lang/queryAst/getSafeInsertIndex.ts

```diff
@@ -0,0 +1,36 @@
+import { getIdentifiersInProgram } from './getIndentifiersInProgram'
+import { Program, Expr } from 'lang/wasm'
+import { Node } from '@rust/kcl-lib/bindings/Node'
+import { getTagDeclaratorsInProgram } from './getTagDeclaratorsInProgram'
+
+/**
+ * Given a target expression, return the body index of the last-used variable
+ * or tag declaration within the provided program.
+ */
+export function getSafeInsertIndex(
+  targetExpr: Node<Program | Expr>,
+  program: Node<Program>
+) {
+  const identifiers = getIdentifiersInProgram(targetExpr)
+  const safeIdentifierIndex = identifiers.reduce((acc, curr) => {
+    const bodyIndex = program.body.findIndex(
+      (a) =>
+        a.type === 'VariableDeclaration' && a.declaration.id?.name === curr.name
+    )
+    return Math.max(acc, bodyIndex + 1)
+  }, 0)
+
+  const tagDeclarators = getTagDeclaratorsInProgram(program)
+  console.log('FRANK tagDeclarators', {
+    identifiers,
+    tagDeclarators,
+    targetExpr,
+  })
+  const safeTagIndex = tagDeclarators.reduce((acc, curr) => {
+    return identifiers.findIndex((a) => a.name === curr.tag.value) === -1
+      ? acc
+      : Math.max(acc, curr.bodyIndex + 1)
+  }, 0)
```

Comment:
Both the `tagsDeclarators` and the identifies `reduce(...findIndex)` i.e. a nested loop, so how long this will take to run is ~the square of the program length. It's not something I'm terribly worried about, so consider it a nit.

But could probably be resolved by using a map/object with the var/tag name being the key instead of an array. P.S. I know I suggested `{ tag: TagDeclarator; bodyIndex: number }[]`, sorry.